db/src/db.rs

```diff
@@ -899,135 +970,216 @@ mod tests {
+                          [101 :db/ident :keyword/value2 ?tx false]
+                          [101 :db/ident :keyword/value22 ?tx true]
+                          [?tx :db/txInstant ?ms ?tx true]]");
+        assert_matches!(conn.datoms(),
+                        "[[100 :db/ident :keyword/value11]
+                          [101 :db/ident :keyword/value22]
+                          [200 :db.schema/attribute 100]
+                          [200 :db.schema/attribute 101]]");
+
+
+        // Test that asserting existing :db.cardinality/one elements doesn't change the store.
+        conn.transact("[[:db/add 100 :db/ident :keyword/value11]
+                        [:db/add 101 :db/ident :keyword/value22]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[?tx :db/txInstant ?ms ?tx true]]");
+        assert_matches!(conn.datoms(),
+                        "[[100 :db/ident :keyword/value11]
+                          [101 :db/ident :keyword/value22]
+                          [200 :db.schema/attribute 100]
+                          [200 :db.schema/attribute 101]]");
+
+
+        // Test that asserting existing :db.cardinality/many elements doesn't change the store.
+        conn.transact("[[:db/add 200 :db.schema/attribute 100]
+                        [:db/add 200 :db.schema/attribute 101]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[?tx :db/txInstant ?ms ?tx true]]");
+        assert_matches!(conn.datoms(),
+                        "[[100 :db/ident :keyword/value11]
+                          [101 :db/ident :keyword/value22]
+                          [200 :db.schema/attribute 100]
+                          [200 :db.schema/attribute 101]]");
     }
 
     #[test]
     fn test_retract() {
-        let mut conn = new_connection("").expect("Couldn't open in-memory db");
-        let mut db = ensure_current_version(&mut conn).unwrap();
-
-        // Does not include :db/txInstant.
-        let datoms = debug::datoms_after(&conn, &db.schema, 0).unwrap();
-        assert_eq!(datoms.0.len(), 88);
-
-        // Includes :db/txInstant.
-        let transactions = debug::transactions_after(&conn, &db.schema, 0).unwrap();
-        assert_eq!(transactions.0.len(), 1);
-        assert_eq!(transactions.0[0].0.len(), 89);
-
-        let value = edn::parse::value(include_str!("../../tx/fixtures/test_retract.edn")).unwrap().without_spans();
-
-        let transactions = value.as_vector().unwrap();
-        assert_transactions(&conn, &mut db.partition_map, &mut db.schema, transactions);
+        let mut conn = TestConn::default();
+
+        // Insert a few :db.cardinality/one elements.
+        conn.transact("[[:db/add 100 :db/ident :keyword/value1]
+                        [:db/add 101 :db/ident :keyword/value2]]").unwrap();
+        assert_matches!(conn.last_transaction(),
+                        "[[100 :db/ident :keyword/value1 ?tx true]
```

Comment:
I'd like to do all of these tests as separate tests, like we do in the Clojure implementation: https://github.com/mozilla/mentat/blob/84a80f40f5c888f8452d07bd15f3b5fba49d3963/test/datomish/db_test.cljc#L426.

Filed https://github.com/mozilla/mentat/issues/368 to track this.  I'll land without it, but I'm happy to get to this fairly quickly, since it's important to get it right.